server.go

```diff
@@ -43,14 +57,91 @@ func NewServer(connBufferSize int, httpPort, grpcPort int) *Server {
 //Init starts the server
 func (s *Server) Init() {
 	go s.listen()
-	s.fetchBigBang()
+	if s.StateFileName != "" {
+		err := s.loadState(filepath.Join(statesFolderName, s.StateFileName))
+		if err != nil {
+			fmt.Println("\nLoading state failed, exiting now", err)
+			panic(err)
+		}
+	} else {
+		s.fetchBigBang()
+	}
 }
 
 //Run offloads the computation of changes to cis
 func (s *Server) Run() {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
 	for {
+		if len(signals) != 0 {
+			s.grpcServer.Stop()
+			fmt.Println("Received Signal:", <-signals)
+			break
+		}
 		s.step()
 	}
+
+	err := s.saveState(time.Now())
+	if err == nil {
+		fmt.Println("\nState successfully saved")
+	} else {
+		fmt.Println("\nState could not be saved:", err)
+	}
+}
+
+func createDirIfNotExist(dir string) error {
+	_, err := os.Stat(filepath.Join(dir))
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0755)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (s *Server) saveState(saveTime time.Time) error {
+	err := createDirIfNotExist(statesFolderName)
+	if err != nil {
+		return err
+	}
+	temporaryPath := s.buildTemporaryStateFilePath(saveTime)
+	file, err := os.Create(temporaryPath)
+	if err != nil {
+		return err
+	}
+	encoder := gob.NewEncoder(file)
+	err = encoder.Encode(s)
+	if err != nil {
+		return err
+	}
+	err = file.Close()
+	if err != nil {
+		return err
+	}
+	return os.Rename(temporaryPath, s.buildStateFilePath(saveTime))
+}
+
+func (s *Server) loadState(statePath string) error {
+	file, err := os.Open(statePath)
+	if err != nil {
+		return err
+	}
+	decoder := gob.NewDecoder(file)
+	err = decoder.Decode(&s)
+	if err != nil {
+		return err
+	}
+	return file.Close()
+}
+
+func (s *Server) buildStateFilePath(saveTime time.Time) string {
+	return filepath.Join(statesFolderName, string(saveTime.Format("20060102150405")))
+}
+
+func (s *Server) buildTemporaryStateFilePath(saveTime time.Time) string {
+	return filepath.Join(statesFolderName, "SAVING_"+string(saveTime.Format("20060102150405")))
```

Comment:
`buildStateFilePath` and `buildTemporaryStateFilePath` are not using anything from the Server, so they should not be defined on `*Server`